Color UI remaining time

mjbogusz opened the pull request:
### Description
Add the possibility of displaying remaining print time with `TFT_COLOR_UI`.

The time remaining logic was copied from DOGM code, as it seemed the most complete - either taking time from M73 or extrapolating time elapsed by print progress.

Additionally the display items (in particular the feed/flow rate) are slightly rearranged to make room for elapsed/remaining time icons.

### Requirements
Requires an LCD display.

### Benefits
Fixes #19780, improves the display usability/readability.

### Configurations
I'm testing this with MKS Robin Nano 1.2 and its 3.5" 480x320 LCD.

The important bits are:
`Configuration.h`:
```cpp
// Line 3125
#define TFT_COLOR_UI
```
`Configuration_adv.h`:
```cpp
// Line 1442
#define SHOW_REMAINING_TIME
```

### Related Issues
Fixes #19780.

### TODO before exiting WIP status:
- [x] Implement for 320x240, 1024x600 displays
- [x] Add icons for time elapsed/remaining
- [ ] Add unit tests
- [x] Rebase on top of #24767 (once it's merged)
- [ ] Fix bugs/errors

mjbogusz commented:
Current status: very much a draft/WIP.
The TODO list to be completed before becoming not-WIP:
* Implement for 320x240, 1024x600 displays
* Add icons for time elapsed/remaining
* Add unit tests
* Fix bugs/errors, obviously

EvilGremlin commented:
See https://github.com/MarlinFirmware/Marlin/pull/24767 - it's going to be merged soon, so better base your code on that.

mjbogusz commented:
> See #24767 - it's going to be merged soon, so better base your code on that.

I've seen it and I'll gladly rebase once it gets merged.
Luckily it doesn't modify the files I'm changing here so I hope it'll be mostly just switching some conditionals around.

---

Right now this PR results in this layout (once with an estimated time remaining available, one without):

![IMG_20221008_045346](https://user-images.githubusercontent.com/4386503/194685796-c57d5d99-b663-43ab-8609-569bf3671d69.jpg)
![IMG_20221008_045404](https://user-images.githubusercontent.com/4386503/194685794-01b9492e-706c-487f-a4b3-ee00776d2066.jpg)

mjbogusz commented:
I've addressed the comment on the commit - if possible, I'd like to ask to comment on code within the PR's context and not the commit's, because as it is now the comment (and response) is not visible in the discussion here.

Tests/builds are broken - as for everyone on `bugfix-2.1.x` - due to the example configs being already updated/merged for #24767 but not the code.
I must say I have at best mixed feelings about such a workflow around breaking changes, but that's a topic for another time and place.

mjbogusz commented:
I've added icons for time elapsed and remaining. For now only the 32px bitmaps and classes were added - is there a place where I could upload the entire GIMP project files (128x128, with layers, paths and all that stuff) without littering the repo?

The UI with the icons looks like this:
![IMG_20221008_211948](https://user-images.githubusercontent.com/4386503/194724332-0565d49d-4366-42f1-b4c9-39828dd64375.jpg)

EvilGremlin commented:
Yeah, it's very annoying when you have to rename things in conf... i vote for ditching example confs dependency and putting everything in tests.
Constants - it's not really a rule, more of a general convenience to keep all compile-time stuff in preprocessor, for ease of maintenance etc. `xy` being exception in color_ui because there's just so much of it.
BTW - here's script for 4bit images  https://github.com/EvilGremlin/Marlin/blob/newcolor_ui/Marlin/src/lcd/tft_newcolor/convert.py

mjbogusz commented:
> Yeah, it's very annoying when you have to rename things in conf... i vote for ditching example confs dependency and putting everything in tests.

Agreed - tests belong with the code and if the example configs are tests, then they belong here too.
Not only should it be easier for config key changes but it would also drop the implicit dependency on internet connection for running tests - I was quite surprised when I've stumbled upon it in the `use_example_config` script.

> Constants - it's not really a rule, more of a general convenience to keep all compile-time stuff in preprocessor, for ease of maintenance etc. `xy` being exception in color_ui because there's just so much of it.

I see the point being to have it solved by the compiler, though I'd assume a single-define-single-use variable should be optimized away anyway. `x`/`y` variables are not and cannot be constant, at least in most of the code I've touched (e.g. `y` specifically is incremented).

> BTW - here's script for 4bit images https://github.com/EvilGremlin/Marlin/blob/newcolor_ui/Marlin/src/lcd/tft_newcolor/convert.py pic source - put in same place they are now, but better redo them in Inkscape

Now this is a script I've been missing! The previous version was a crude modification of the [bmp2hex](https://github.com/robertgallup/python-bmp2hex) script, but this will save me some time. I think it'd be great for future contributors if it was available in `buildroot/bin` with at least a short description/mention in the contribution guide.

I'll redo the pics in Inkscape and correct the "shadow" while at that as it was slightly inconsistent with the existing ones.
Are there by any chance sources of the existing icons available anywhere?

mjbogusz commented:
I've added new icons redone in Inkscape, including the svg project files.

The conversion script was... a bit rough on the edges - I've modified it quite a bit: https://github.com/mjbogusz/Marlin/blob/bitmap2cpp/buildroot/bin/bitmap2cpp

mjbogusz commented:
I've rebased onto the updated bugfix branch (after #24767 merge) and updated the code to utilize the new flags.

EvilGremlin commented:
:+1: 
i feel bitmap2cpp is way too generic name for what it is

mjbogusz commented:
320x240 version added, however I have no such display to test it.

EvilGremlin commented:
just use simulator...

mjbogusz commented:
Oooh that'll help for sure, thanks.

Wish it was described in the [development docs](https://marlinfw.org/meta/development/) - I'm definietely gonna put a PR for the documentation after finishing this PR with this and my other findings.
The simulator in particular is effectively undiscoverable - after reading your message I duckduckwent for "marlin firmware simulator" and it turned up nothing and google only found a single Twitter thread.

Anyway, back to working on the 1024x600 version.

EvilGremlin commented:
native.ini have almost all info you need

mjbogusz commented:
Rebased onto updated `bugfix-2.1.x` again.

Quick notes on setting up the simulator on WSL before I forget:
* Setup [VcXsrv](https://sourceforge.net/projects/vcxsrv/), add required things into your bashrc/zshrc
* `source ~/.platformio/penv/bin/activate`
* `buildroot/bin/use_example_configs "Simulator"`
* `sudo apt install libglm-dev libsdl2-dev libsdl2-net-dev`
* `export LIBGL_ALWAYS_INDIRECT=0`
* open `.pio/libdeps/simulator_linux_debug/MarlinSimUI/src/MarlinSimulator/window.cpp`, comment out lines 39-40 (setting multisampling attributes); if the file doesn't exist try building the binary once to let PIO download the deps
* `platformio run --environment simulator_linux_debug`
    * or, if building via VSCode:
        * open the PlatformIO environment selector and select `simulator_linux_debug` or `simulator_linux_release` (NOT `linux_native`!)
        * build
* `.pio/build/simulator_linux_debug/MarlinSimulator` should launch the simulator!

Now, for simulating/testing the ColorUI:
* enable `TFT_COLOR_UI`, disable `REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER` in `Configuration.h` (lines 45/48)
* [optionally] adjust the resolution in lines 3144-3147 of `Configuration.h`
* apply some random patches:
    * define `TOUCH_MISO_PIN`, `TOUCH_MOSI_PIN`, `TOUCH_SCK_PIN` in `Marlin/src/HAL/NATIVE_SIM/spi_pins.h`
    * remove/comment out `void MarlinUI::rotate_progress()` in `Marlin/src/lcd/marlinui.cpp` (lines 1731-1745)
* For touch screen buttons:
    * enable `TOUCH_SCREEN` (`Configuration.h`, line 3194)
    * disable `TOUCH_SCREEN_CALIBRATION` (`Configuration.h`, line 3202)

I'll try to open PRs with the fixes/adjustments mentioned above, but first I want to finish this one - it proved to be much bigger than I thought it'll be ;)

mjbogusz commented:
320x240 layout with everything enabled:
![image](https://user-images.githubusercontent.com/4386503/195186354-92a78c73-e0b7-40a6-bd0d-2ff6134f28af.png)

mjbogusz commented:
> Minor critique: I think a dash like `-` might be better than `N/A` or just hide it if the printer is idle.

I'll change it to dash, IMO hiding it would make the UI sort of 'unbalanced'. I've also noticed I have to adjust the top margins for the time elapsed/remaining.

thisiskeithb commented:
> it would make the UI sort of 'unbalanced'

Wait until you see the 1024x600 UI used on the BX! Hah!

![image](https://user-images.githubusercontent.com/13375512/195187826-7f86367d-f6b4-45b5-92cf-7df6ae4facc1.png)

mjbogusz commented:
I've meant that there would be 2 items in the upper row and only one on the side of the lower one.
1024x600 is next, here we go! ;)

320x240 after adjustments:
![image](https://user-images.githubusercontent.com/4386503/195188296-6c642725-498c-4763-86c1-795151402cb5.png)

480x272:
![image](https://user-images.githubusercontent.com/4386503/195191990-4b2ed02a-3ca0-4a84-a9b3-b9c5ebbf0f14.png)

480x320:
![image](https://user-images.githubusercontent.com/4386503/195191775-74bc2856-31ff-4aa9-949c-8e63c07e0040.png)

mjbogusz commented:
I see what you've meant with the 1024, it's free real estate :D Enlarging the font size, icons, and progress bar would help with readability a lot.

initial layout with times elapsed/remaining (will push soon):
![image](https://user-images.githubusercontent.com/4386503/195196546-c17c3ea6-43e1-484c-aac7-b6934357de40.png)

mjbogusz commented:
The PR is mostly ready, I just need to add some tests - any hints as for where/how to put them? I haven't found an obvious place...
I'm thinking some simple unit tests for `duration_t::toString` and some configuration tests for combinations of different screen sizes, `SHOW_ELAPSED_TIME`, and `SHOW_REMAINING_TIME`.

In the meantime I'm preparing few "side-quest" PRs with some minor fixes/changes I had to make to test the ColorUI in simulation (especially the 1024x600 was problematic), one already pushed to MarlinSimUI and linked above.

rq3 commented:
> The PR is mostly ready, I just need to add some tests - any hints as for where/how to put them? I haven't found an obvious place... I'm thinking some simple unit tests for `duration_t::toString` and some configuration tests for combinations of different screen sizes, `SHOW_ELAPSED_TIME`, and `SHOW_REMAINING_TIME`.
> 
> In the meantime I'm preparing few "side-quest" PRs with some minor fixes/changes I had to make to test the ColorUI in simulation (especially the 1024x600 was problematic), one already pushed to MarlinSimUI and linked above.

Thos works extremely well on my 480x320 Anycubic display, I thank you very much for the effort, and it closes a feature request I made long ago.

Attached is a small collection of files (as long as your're playing with the 480x320.cpp). These make the progress bar equal in size to the axis bar, makes the progress bar vivid green for better contrast, and overall just makes for a more visually pleasing color_ui display. At least for me o my 480x320
[Progress_Bar.zip](https://github.com/MarlinFirmware/Marlin/files/9795834/Progress_Bar.zip)
.

thisiskeithb commented:
> makes the progress bar vivid green for better contrast, and overall just makes for a more visually pleasing

Changing the background color to black also helps a lot.

mjbogusz commented:
> This works extremely well on my 480x320 Anycubic display, I thank you very much for the effort, and it closes a feature request I made long ago.

I see no icons next to the times on your photo - is this an early version?

> Attached is a small collection of files (as long as your're playing with the 480x320.cpp). These make the progress bar equal in size to the axis bar, makes the progress bar vivid green for better contrast, and overall just makes for a more visually pleasing color_ui display. And it doesn't seem to interfere with your effort, at least for me on my 480x320 display. 

This seems like a great follow-up PR!

rq3 commented:
> > This works extremely well on my 480x320 Anycubic display, I thank you very much for the effort, and it closes a feature request I made long ago.
> 
> I see no icons next to the times on your photo - is this an early version?
> 
> > Attached is a small collection of files (as long as your're playing with the 480x320.cpp). These make the progress bar equal in size to the axis bar, makes the progress bar vivid green for better contrast, and overall just makes for a more visually pleasing color_ui display. And it doesn't seem to interfere with your effort, at least for me on my 480x320 display.
> 
> This seems like a great follow-up PR!

I'm not sure where the icons went. I probably screwed up the hand editing of the requisite files. Either that, or the wider progress bar pushed them out? Probably the former, knowing me!

rq3 commented:
> > Minor critique: I think a dash like `-` might be better than `N/A` or just hide it if the printer is idle.
> 
> I'll change it to dash, IMO hiding it would make the UI sort of 'unbalanced'. I've also noticed I have to adjust the top margins for the time elapsed/remaining.

I don't care for the dash. It looks like flakey pixel on the display, but that's just me.

mjbogusz commented:
Rebased and un-drafted, still waiting for any feedback regarding unit tests but I think it's ready for comments and review otherwise.

As for the time remaining while not printing: I wouldn't use a question mark - for the XYZ values I understand it as "unknown / be cautious with movement commands", while here we're trying to convey a "not applicable" message.

rq3 commented:
> Rebased and un-drafted, still waiting for any feedback regarding unit tests but I think it's ready for comments and review otherwise.
> 
> As for the time remaining while not printing: I wouldn't use a question mark - for the XYZ values I understand it as "unknown / be cautious with movement commands", while here we're trying to convey a "not applicable" message.

The question mark is purely aesthetics for me. I see it as a "waiting for valid data" symbol (like the axes), but to each their own.

Your remaining time code works fantastically well on my delta with an STM32F1 board with 480x320 TFT touch screen. Many thanks! Nice job

mjbogusz commented:
@thisiskeithb sorry for the ping but the PR still waits for feedback. Lemme know if you want me to rebase before review.
By the way- why is the related issue closed&locked? It is NOT solved as this PR is, well, not merged...

thisiskeithb commented:
> why is the related issue closed&locked

We close issues when PRs are created. It's just how things are done around here.

> Sorry for the ping but the PR still waits for feedback

Only thinkyhead can merge code, so we'll need to wait for his review/merge.

mjbogusz commented:
> We close issues when PRs are created. It's just how things are done around here.

Fair enough - it was just unusual. I see the potential race condition of closing an issue and then dropping a PR due to, e.g., bad code quality, but I guess on the other hand it helps to keep the issue tracker cleaner.

> Only thinkyhead can merge code, so we'll need to wait for his review/merge.

Should I rebase and ping him? Or just be patient and wait for a merge window, linux-style?

EvilGremlin commented:
Yep and yup. Sometimes GH glitch and Scoot can't get PR up to date himself.

quiret commented:
I love improvements to the Color UI! Did you know that thinkyhead is pushing for everyone to try this new amazing Marlin UI? He even says to prefer it over FSMC because it exposes Marlin more natively. I hugely agree with him! Want to hear from the man himself? Check out this video series (part 1):

https://www.youtube.com/watch?v=9ib57X7dhJ0

EvilGremlin commented:
What you mean "prefer it over FSMC"? You make no sense.

quiret commented:
> What you mean "prefer it over FSMC"? You make no sense.

Do you know those DWIN displays? That's what I meant. FSMC may have been the wrong technical word! But yes, more power to the color UI ;)

EvilGremlin commented:
Not even close... DWIN is HMI display and ColorUI is drawn 100% natively (using FSMC parallel or SPI interface)

EvilGremlin commented:
It's becoming a burden... I don't like existence of extUI as a separate instance, i think it must become the only Marlin's UI API. It'll make everything much simpler. But it's be big work to rewrite/refactor all native UIs, and footprint must be kept the same...

quiret commented:
> We don't have a choice on what displays/hardware/firmware vendors ship their printers with (or what users buy) so improving code/features where we can, benefits users of said hardware.
> 
> Running MarlinUI on whatever enabled display hardware will always give you access to the most menus & features since you don't have to rewrite code to fit a specific UI (`DWIN_CREALITY_LCD`, `DWIN_LCD_PROUI`, `DWIN_CREALITY_LCD_JYERSUI`, `TFT_LVGL_UI`, `TOUCH_UI_FTDI_EVE`, etc.)

I love this community! You realize TMC Steppers have an option in Color UI. Then a guy comes along and adds the option into DWIN (pretty recent commit). Also I have been in contact with some hardware vendors and they are very excited about the work that is coming to Marlin, especially my stability improvements to the SPI! I tell you there is big potential in every aspect in Marlin and good promotion is key.

> It's becoming a burden... I don't like existence of extUI as a separate instance, i think it must become the only Marlin's UI API. It'll make everything much simpler. But it's be big work to refactor all native UIs, and footprint must be kept the same...

Got it. I know how you feel because I realized similar burden in code structures. I want more documentation about Marlin internals, maybe in a video series or a picture-ful blog. Somebody has got to compare and judge things, make an open inquiry.  But you got to have the maintainers too, and I would not want to look over development redundancies if there is a better one around! You say extUI... Might look into it next year... I mean I would be all for a refactoring of Marlin...

rq3 commented:
What can I do to help get this PR merged?

mjbogusz commented:
Rebased - please re-test, as merging some of the changes in the `ui_[resolution].cpp` was quite tricky.

This is ready for review (and merge if deemed good enough) - @thinkyhead if you could (:

rq3 commented:
> Rebased - please re-test, as merging some of the changes in the `ui_[resolution].cpp` was quite tricky.
> 
> This is ready for review (and merge if deemed good enough) - @thinkyhead if you could (:

I am looking forward to seeing this merged. I have been running this on a hand edited version of bugfix2.1.2, but doing the same for the latest 2.1.3 bugfix is problematic.

Anycubic 3.5 TFT, 480x320.

mjbogusz commented:
I'm trying to rebase AGAIN, but https://github.com/MarlinFirmware/Marlin/pull/25207 is throwing me in for a loop with lots of conflicts.
I wish it was done as a separate UI class/file because the code is getting quite spaghettified and IMO vertical screens call for different UI layouts than horizontal ones.

I think I'll try to make it more modular/manageable as a follow-up, maybe dividing the screen into sections and handling them off to separate methods or something along these lines.

Hopefully I'll push the rebased version soon, but I'm not ruling out disabling the time remaining code for portrait orientation.

mjbogusz commented:
It's up and SHOULD work even on the portrait mode 240x320 displays - @alexgrach sorry to ping you but could you test and verify as you were the one that introduced this specific configuration?

@thinkyhead RFC again (:

alexgrach commented:
Should I check this [branch](https://github.com/mjbogusz/Marlin/tree/color_ui_remaining_time)?

mjbogusz commented:
Yes - if you have a working repo clone you can just add my fork as an additional remote and then checkout the branch directly, just remember to backup your printer configuration and to enable `SHOW_ELAPSED_TIME` and `SHOW_REMAINING_TIME` for the tests.

alexgrach commented:
You can check it in simulator I think.

```
// This should work
#define TFT_RES_320x240
#define TFT_ROTATION TFT_ROTATE_270
```
![UI](https://user-images.githubusercontent.com/7253483/215591328-0d811b69-a740-414c-981d-2ecfd6556204.JPG)

mjbogusz commented:
I didn't have my full setup at hand, I'll try to set it up again and test it in the coming days. However last time I was testing I had to significantly modify the simulator setup to test all the previously "supported" resolutions, hopefully it won't be so hard this time around ;)

OneMonkeyArmy commented:
I have written this GUI from the ground up, starting by the driver for the TFT, then to accelerate the drawing of any image or pixel to TFT buffer , I wrote my own graphic library that draw directly into the ram buffer... DMA is taking care of the refresh... 
All the Fonts management is custom, and it manage multiple fonts. with Scale X and Y, BOLD, as I write it into the ram buffer, even developed a spatial antialiasing automatically applied when the scale of the fonts is larger than 100% , and all is done in real time...

I also added a basic 3D library, to display the bed levelling topography of the bed, but this is under development..
![image](https://user-images.githubusercontent.com/84219749/219875392-eb8b414e-d3f3-4cdd-92e9-966c75553506.png)

It will some time and effort to make it fully compatible with marlin, but if we do it, because I won't do it alone !  It can only run on Boards equipped with at least a STGM32F407 with a TFT connected directly to the board with 16 bit FSMC with DMA, then you get power and responsiveness... All the Display these days are just a bidirectional serial remote control, in which another firmware is running on a 2nd MCU...  


![328530215_1363771274477604_8415890372917922299_n - Copy](https://user-images.githubusercontent.com/84219749/219873055-64db5e77-fa55-4ccf-bd29-1557f49e5044.jpg)
![328660136_943691020145084_8109315832220348005_n (1)](https://user-images.githubusercontent.com/84219749/219873057-aab8f4dd-ccc5-47cb-95d5-a31e8de40f0c.jpg)
![328660136_943691020145084_8109315832220348005_n](https://user-images.githubusercontent.com/84219749/219873062-48e6a13c-410e-4994-afd4-400192c9e9bd.jpg)
![328739442_5916467835102188_1875505271479897863_n (1)](https://user-images.githubusercontent.com/84219749/219873063-5d3305f0-5bfa-4c3a-b43b-8ccc0d01096a.jpg)
![328942327_1641764506283617_5011250660249703085_n](https://user-images.githubusercontent.com/84219749/219873064-05dc67f1-a86b-4fcf-9968-51cb28435068.jpg)
![328973499_603394771611417_226623739862042443_n (1)](https://user-images.githubusercontent.com/84219749/219873065-606e6a9e-ab75-4b9a-aee9-0c8602d8c1e9.jpg)
![329009127_1235060214105830_5738755292175223059_n](https://user-images.githubusercontent.com/84219749/219873066-6a45b27f-118d-4c68-8e97-eaa3776c5214.jpg)
![329155426_886970845849477_5130650208229603495_n](https://user-images.githubusercontent.com/84219749/219873067-d154895a-08f6-46a1-b977-59951712826f.jpg)
![329166035_701698405028952_2999026982617544618_n](https://user-images.githubusercontent.com/84219749/219873069-a46ef674-29b7-42ce-b71c-9bd2cf70b8c4.jpg)
![329166472_3312818422306851_5049448331508332452_n](https://user-images.githubusercontent.com/84219749/219873070-63a7d73b-bb96-4f6f-a5fd-51c4d29f84c7.jpg)
![329234164_748851366773349_6204146768752293362_n](https://user-images.githubusercontent.com/84219749/219873071-5087dd17-9b70-4f31-ba1c-75e018339f85.jpg)
![329329364_566827865164685_4809918846046548705_n (1)](https://user-images.githubusercontent.com/84219749/219873072-b1a6c697-d3a5-4a3e-8151-ed60048f6054.jpg)
![329739775_594442468733024_8741888370112134018_n](https://user-images.githubusercontent.com/84219749/219873074-38bd1a09-12ba-411e-b0f2-fef82ec19cd1.jpg)
![329815884_452835273636052_9126317282442775930_n](https://user-images.githubusercontent.com/84219749/219873075-7c2e4d0e-ec9b-463c-bfc4-1dc989822da6.jpg)
![329850386_2343704822473546_1584178989575440028_n](https://user-images.githubusercontent.com/84219749/219873076-52d1c4aa-d0e7-4d39-b5ca-dfef009d45ce.jpg)
![323429613_572442294905344_3830503737311866871_n (1)](https://user-images.githubusercontent.com/84219749/219873077-8c24eb2e-5ebb-4fa3-861c-7ca1b5339e48.jpg)

OneMonkeyArmy commented:
I never fully understand why Marlin was hard to maintain, until I read it completely, and start to experiment with it..  And to make my own GUI, I had to remove everything related to any other display, it is a non sens to try to have only one version that runs on all MCU AVR or ARM 8 bits or 32 bits, with a serial display B/W from the 90`s or a Super fast 16 bit bus full HD, and for all kind of printers, Delta, Core, Cartesian, polar...   And why the EEPROM size is 4K, no printer uses 1/2 of it, because when I use a Delta, I do not need to store the Core XY settings, but the space is reserved !!  BTW..  a byte can contain 8 bools... !!! So my conclusion is there should be a few versions of marlin , I do not know why, every time I modify something, I have to read what is used by my system and what that is not, I was always loosing my time with the AVR, the 16Mhz exception, the 4 row LCD, the option that is not on my system and will never be.. Not that it will change something once the compiler manage all the condition.. the firmware will take the same space, but for the use of mods and updates, it is so much easier to read and understand... ( and finally wrote my own firmware dedicated for the STM32F407, but it is a lot faster to compile, edit and knowing every line in it!  
I read long time ago, a few lines from a contributor of the past !!! That the use of  #DECLARE and macros should be top priority, to save space and MCU clock cycle... That was true before, but now we have incredibly fast MCU with huge flash memory.
I think that part of this suggestion should be adapted... for instance, using "COMPILER DECLARE, CONDITIONS, AND MACROS" to manage standard stepper driver pins like ENABLE, STEP AND DIR... prevent the possibility of linking all the ENABLE pin of X Y Z axis together, and stopping the useless dance of enable X, ENABLE Y, ENABLE Z one by one... of course, I can, but I have to Recompile... My new approach, does not rely on 2 huge configuration files, that user access, modify, and compile the Firmware until he obtains what he wants... it can be done either by providing the firmware an XML file, or for simple changes, just on the GUI itself... 

This is what Repetier server receive when marlin reports temp and PWM:

T:25.60 /0.00 B:25.00 /0.00 C:25.00 /0.00 T0:25.60 /0.00 T1:25.10 /0.00 @:0 B@:0 C@:0 @0:0 @1:0...

This can be easily shortened by a factor of 2.. And at every second, or according to report delay, when resent, why the unchanged values are still in the line transmitted... And lastly but not the least, why the @ used as identification of the PWM value of a specific heater, is for the BED B and enclosure C, placed after the letter @B: and @C:, but for the others it switches position...it is after?  0@: 1@:...   like Huell in breaking bad... I will conclude with "all I am saying !"

EvilGremlin commented:
Ehmm... it's all very interesting and good job, but why in this thread? Anyway, your opinions were voiced many times before, but for now it's all idle thoughts shelfed for eventual marlin v3.

OneMonkeyArmy commented:
  18.02.2023, 12:44, "EvilGremlin" ***@***.***>: YOU WROTE ME... "Ehmm... it's all very interesting and good job, but why in this thread?" Read the thread completely, it will all make sens after... Then you continue with.. "Anyway, your opinions were voiced many times before, but for now it's all idle thoughts shelfed for eventual marlin v3." When I comment on specific things, it is because I have already done it in my firmware, so.. no, it is not idle thoughts shelfed for eventual marlin v3... It is my humble POV, and experience I have done in my firmware that, by the way, is no longer related to Marlin... it is a completely new Firmware, with a compatibility with others at 80%, and many different and unique things.. so I am working with Ultimaker and Repetier, to add my Firmware as a new one, with it's specific characteristic, so my systems will be compatible with them... Everyday, I receive emails and sms from all over the world... for details about this new firmware... And every day, I speak, negotiate, or explain things... With many Leaders in this industry...  I do not just have Ideas shelfed, I make things change by my own will... my ideas shelfed, are top secret...The Lead Engeineer at Dremel is Mike Crowe, we talked so much by phone and SMS since a year....The Microcontroller Marketting Manager at STMicroelectronics, Tilen Majerle. and he always in a rush to answer me...I have done major changes in the Klipper configuration file, to help developers of Klippers to make all board powered by STMicroelectronics compatible with Klipper... I did NOVEMBER 22, AND JUST SAW THAT VORON AND VZ-BOT, USES MY IMPROVED VERSION...I answer to many persons around the world everyday, for free, just to help...When I make phone calls in Germany at the head office of Trinamic, They know me, and always take time for me, of course I asked them a quote for 10000 x TMC2209, and the delay each time we will order them..and also wanted to know how long they will make this particular driver.. he answered me, as long as you going to buy them..anyway... To conclude, before saying something to a stranger, google its name, you may change your comment to something different... Recently I was developing the new controller board for Dremel Digilab, but we didn't agree on the contract, I spent a month for nothing, but, I went back to my firmware project..  Daniel Massicotte, AKA "The One monkey army"  Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you commented.Message ID: ***@***.***>

EvilGremlin commented:
Wowowow, easier with flex :smiley:  I meant that whatever you say isn't happening in any forseeable future, not in marlin 2.x. And i still don't get you, did you choose random thread to promote your new firmware or what? Then where are links? I don't see how anything you write is any relevant to this PR.

rq3 commented:
> I didn't have my full setup at hand, I'll try to set it up again and test it in the coming days. However last time I was testing I had to significantly modify the simulator setup to test all the previously "supported" resolutions, hopefully it won't be so hard this time around ;)

@mjbogusz , attached is what I am currently running on my 480x320 TFT, thanks to your fantastic effort!
The changes are:
1) Pushed the SD card and Settings icons as far outboard as possible to allow room for the remaining time display. Not perfect, there is still minor clipping, but it's almost invisible unless you are looking for it when the displayed digits are many. I tried making the SD and Settings icons 32 pixels, but it looked nasty.
2) Made the progress bar equal in height to the axis bar for visual symmetry, and changed its color to vivid green for contrast.
3) Very minor re-alignment (a few pixels) for feed, flow, elapsed, and remaining...just to make things pretty (to my eye).
4) Used a "?" for remaining time when no data is available, in keeping with the axis data. Just a personal preference.
5 Adjusted vertical position of status data (Bed Heating..., etc.) for visual symmetry.

Thank you!
[Remaining_Time.zip](https://github.com/MarlinFirmware/Marlin/files/10845278/Remaining_Time.zip)
![IMG_1901](https://user-images.githubusercontent.com/3343927/221716325-2eddbe07-392c-4479-9574-596ee00fc14d.jpg)
![IMG_1900](https://user-images.githubusercontent.com/3343927/221716328-cde0529b-3db8-424e-9e65-7987373dbbb1.jpg)

thinkyhead commented:
> it is a non sens to try to have only one version that runs on all MCU AVR or ARM 8 bits or 32 bits, with a serial display B/W from the 90`s or a Super fast 16 bit bus full HD, and for all kind of printers, Delta, Core, Cartesian, polar

@OneMonkeyArmy —

You may disagree with the philosophy behind Marlin, but it is what it is. Marlin has a continuity of development from the beginning in 2011 to today, and we are simply continuing to build upon it. Maintaining this code takes up all the time that I might otherwise have to start over from scratch.

I respectfully disagree that we should abandon older hardware from the ancient past of 10 years ago. There are still millions of machines using that hardware, including many of our own. There is simply no way we can leave those behind, and we also don't want to. Marlin is for hobbyists to tinker with, and we want those tinkerers to be able to mix-and-match whatever hardware they have in their stockpile.

It sounds like you prefer something else, and that is fine. RepRapFirmware (for example) makes a very good starting point for an all-new 32-bit-only firmware. So carry on, and have fun with your projects!

thinkyhead commented:
> Remaining_Time.zip contains the changed files I am using, if they are of any help.

@rq3 — The best way to contribute changes is to submit a PR to the branch that this PR is based upon, [mjbogusz:color_ui_remaining_time](https://github.com/mjbogusz/Marlin/tree/color_ui_remaining_time), and then when mjbogusz merges your PR into his branch it will automatically update this PR.

vovodroid commented:
I've tried this PR on latest Marlin. It shows very strange times. I have 5 hours print containing M73 as follows:

```
Line 12: M73 P0 R317
Line 57: M73 P0 R316
Line 1030: M73 P0 R315
Line 2762: M73 P0 R314
Line 4506: M73 P0 R313
Line 4538: M73 P1 R313
Line 6416: M73 P1 R312
Line 8219: M73 P1 R311
Line 9520: M73 P1 R310
Line 9859: M73 P2 R310
```

But screen shows strange huge  remaining time, decreasing slowly:
![image](https://user-images.githubusercontent.com/8691781/229301201-eaa91055-4f0c-4847-9dbb-77b06e268de6.png)
![image](https://user-images.githubusercontent.com/8691781/229301237-b9057872-82d2-4167-a78f-3c9177283257.png)
![image](https://user-images.githubusercontent.com/8691781/229301242-6db3c9f1-d9e6-4e7f-a74b-dbc56d793e28.png)

mjbogusz commented:
Oooh you've found a bug!

In my code I've used `USE_M73_REMAINING_TIME`, which was changed to `SET_REMAINING_TIME` (I believe in the timeframe between opening this PR and now).
Therefore the code doesn't really utilize M73 and only estimates the remaining time by extrapolating time spent by job progress - and as during the heatup the progress is "stuck" on something like 1% or 0.1% (the job didn't really start yet), the extrapolated time remaining can be ridiculously long but will come down to more realistic values as the print progresses.

vovodroid commented:
It's still showing strange huge times going back and forth in wide range, though M73 is the first command in gcode file.

vovodroid commented:
By a way, R parameter is ulong, so there is no meaning in calculation and displaying seconds.

vovodroid commented:
You also use LCD_SET_PROGRESS_MANUALLY which is now SET_PROGRESS_MANUALLY, see Sanity Check:

```
#elif defined(LCD_SET_PROGRESS_MANUALLY)
  #error "LCD_SET_PROGRESS_MANUALLY is now SET_PROGRESS_MANUALLY."

```

vovodroid commented:
Now it's almost good with SET_PROGRESS_MANUALLY:
![image](https://user-images.githubusercontent.com/8691781/230071086-55990af5-fa95-425c-bb6b-a830663f4379.png)

![image](https://user-images.githubusercontent.com/8691781/230071108-ae4fbd93-9525-4be2-a31b-7802eb30e040.png)


Still starts with strange value:
![image](https://user-images.githubusercontent.com/8691781/230071180-66fb7b69-1902-40c7-bf8a-1ffae28d41e4.png)

and gradually arrives to correct value.

vovodroid commented:
Hi,
please replace **LCD_SET_PROGRESS_MANUALLY** with **SET_PROGRESS_MANUALLY**.

thinkyhead committed:
TFT Elapsed additions

Co-Authored-By: mjbogusz <mjbogusz@users.noreply.github.com>

thinkyhead committed:
cosmetic code cleanup

cosmo

Co-Authored-By: mjbogusz <mjbogusz@users.noreply.github.com>

cosmo

thinkyhead committed:
Shift some elements

Co-Authored-By: mjbogusz <mjbogusz@users.noreply.github.com>

thinkyhead committed:
Elapsed Time code

Co-Authored-By: mjbogusz <mjbogusz@users.noreply.github.com>

thinkyhead committed:
misc. obvious updates

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/24843

thinkyhead committed:
post merge

thinkyhead commented:
Hmm I thought perhaps some of these changes were a little strange. We'll need to revert some of them to deal with layout breakage. Rather than revert all changes I'm sure we can just cherry-pick to fix these.

![20230623_215656](https://github.com/MarlinFirmware/Marlin/assets/698003/60efb3b1-30ad-4c54-aaf1-f7a6bb0c3314)

CC: @mjbogusz / @uncellon

uncellon commented:
- missed settings and SD card button for touch screen
- with `SHOW_REMAINING_TIME` the second image shows that new label doesn't fits on the screen

![IMG_20230625_094129](https://github.com/MarlinFirmware/Marlin/assets/23581805/d39a3b0c-637c-4d67-bd73-3b657b574b9d)
![IMG_20230625_094352](https://github.com/MarlinFirmware/Marlin/assets/23581805/20dad49b-e3d8-406c-ba69-1ea1c1238e0a)

uncellon commented:
I checked the code briefly, the settings buttons and the SD card are simply cut out in the `ui_480_320.cpp` file.

ellensp commented:
@uncellon provided configs seem to be firmware stock configs, please try again

uncellon commented:
@ellensp, sorry, I didn't notice that I attached the wrong archive. I updated the post and attached the correct archive

thinkyhead commented:
Thanks for having a look at the layout issues and working towards a patch. I've been all over the place lately with issues of many different flavors. I was going to check the state of the simulator and see how well it handles these different sizes of TFT UI since that would make debugging and fixing these kinds of things go a lot quicker.

lukasradek commented:
Same here... I am using landscape 320x240 and the `SD card` and `Settings` buttons are missing. That includes the touch control areas.

ellensp commented:
add_control(SETTINGS_X, y, menu_main, imgSettings); is missing form 480x320 so no setting icon

vovodroid commented:
Hi, do we really need show seconds? They just take screen space and switching between one and two digits jerking the whole time string.

mjbogusz commented:
Good point.
I think seconds are only relevant in the last stages of a print, so it would make sense to skip them if there's more than an hour left (or elapsed). Similarly, to avoid switching number of digits/characters displayed perhaps the minutes number should be displayed all the time, padded with zeros.

How do you feel about displaying `01h23m` if the time is more than an hour and `00m59s` when it's less?

lukasradek commented:
Seconds are irrelevant for 3D printing regardless of the time left. Since there is so much preprocess / postprocess to be done and prints usually take hours rather than units of minutes, it doesn't really matter if you arrive +- a minute at your printer. 😁

ellensp commented:
I disagree, some of us are impatient and want to see the print finish and want to see the seconds count down.

vovodroid commented:
> some of us are impatient and want to see the print finish and want to see the seconds count down.

Ok, we can show seconds only for last minutes. But I don't see too much value for seconds.

vovodroid commented:
> How do you feel about displaying `01h23m` if the time is more than an hour and `00m59s` when it's less?

I feel very good)))

mjbogusz commented:
I mostly agree; however for example I just love watching the finishing touches on the last layer or two ;) speedprinting people would probably prefer the seconds too - however I must acknowledge that both of these are minor edge cases.

Displaying `01m58s` instead of `00h01m` is just, in my opinion, a bit more elegant.
Also, even if the slicer didn't insert M73, the estimate should be fairly precise towards the end of a print.

vovodroid commented:
Is https://github.com/MarlinFirmware/Marlin/blob/bugfix-2.1.x/Marlin/src/libs/duration_t.h#L130 the only function involved in remaining time? If it also affects elapsed time, I don't think it would be a problem.

mjbogusz commented:
This PR used `toCompactString()` for the remaining/elapsed time if both were enabled - it was added to avoid breaking other places where `toString()` was used. This has changed later when the separate per resolution files were merged and cleaned up in #26077.

The implementation of `toCompactString()` could be replaced with the "two most significant units" as described above and the method used for time elapsed (https://github.com/MarlinFirmware/Marlin/blob/bugfix-2.1.x/Marlin/src/lcd/tft/ui_color_ui.cpp#L339) and remaining (https://github.com/MarlinFirmware/Marlin/blob/bugfix-2.1.x/Marlin/src/lcd/tft/ui_color_ui.cpp#L365)
